src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,98 @@ internal partial class Binder
             { WasCompilerGenerated = wasCompilerGenerated };
         }
 
+        private BoundExpression CreateImplicitNewConversion(SyntaxNode syntax, BoundExpression source, Conversion conversion, bool isCast, TypeSymbol destination, DiagnosticBag diagnostics)
+        {
+            var node = (ObjectCreationExpressionSyntax)source.Syntax;
+            var arguments = ((UnboundObjectCreationExpression)source).AnalyzedArguments;
+
+            TypeSymbol type = destination.StrippedType();
+            BoundObjectInitializerExpressionBase boundInitializerOpt = node.Initializer != null
+                ? BindInitializerExpression(syntax: node.Initializer, type: type, typeSyntax: syntax, diagnostics)
+                : null;
+
+            if (ReportBadTargetType(syntax, type, diagnostics))
+            {
+                return MakeBadExpressionForObjectCreation(node, destination, boundInitializerOpt, arguments);
+            }
+
+            BoundExpression result;
+            switch (type.TypeKind)
+            {
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                    result = BindClassCreationExpression(
+                        node,
+                        typeName: type.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)type,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    break;
+
+                case TypeKind.TypeParameter:
+                    result = BindTypeParameterCreationExpression(
+                        node,
+                        typeParameter: (TypeParameterSymbol)type,
+                        arguments,
+                        boundInitializerOpt,
+                        diagnostics);
+                    break;
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(type.Kind);
+            }
+
+            if (destination.IsNullableType())
```

Comment:
May be good to add a comment on the nullable case to clarify why this is needed: `S? x = new();` means `S? x = new S();`